### PR TITLE
grpc_json_transcoder: Pass through requests when methods not found

### PR DESCRIPTION
**Pull request by tangyan**
Commit Message:

strict_http_request_validation (commit 6ce641f) introduced an
unexpected behavior when the option is turned on.

Before the change (commit 6ce641f), if an url cannot be
matched/recognized by the grpc_json_transcoder, it will pass
through to the filters in the chain, for example, http routers.
This allows the http router to continue to route the request to
other clusters.

After the change (commit 6ce641f), any unmatched/recognized url
will cause a 400 BAD REQUEST.  This is unexpected behavior.

The fix is, if the url can be matched by the transcoder, it means
it is inteneded to be a gRPC transcoding request.  Otherwise, it
should just pass through because at that moment, it's unclear to
the transcoder logic if this unmatched url is supposed to be a
gRPC method or not.

The param validation/resolve continues once the method is matched.

Additional Description:
Risk Level: Low.
Testing: Manual and integration.
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/master/api/review_checklist.md):]

### CODE REVIEW

**Comment by qiwzhang** [comment]
We need a config for such behaviors.  Other people may not like it. 

My suggestion is: since `strict_http_request_validation` was just added a few days ago, assume nobody is using it. we can deprecated it and replace it with behaviors for each error type:

```
error type:  fail or pass_through

e.g.
path_mismatch:  pass_through
unknown_query_parameter: fail
```

Defaults are all pass_through.

**Comment by qiwzhang** [comment]
@nareddyt what do you think?

**Comment by tangyan** [comment]
> We need a config for such behaviors. Other people may not like it.

My patch actually reverts the behavior to be more similar to before the patch 6ce641f.

It's fairly clear to me that: when an unknown url (not the param) is seen by the json grpc transcoder, at that moment, it cannot tell if it is an incorrect url which intended to be transcoded or a correct url which should be passed through.

It is very common (at least to me) that we need to support this because the grpc transcoder actually has already been limited to the set of services in the config.

Here is the example from grpc json transcoder doc:

Proto:

// The greeting service definition.
service Greeter {
  // Sends a greeting
  rpc SayHello(HelloRequest) returns (HelloReply) {
    option (google.api.http) = {
      get: "/say"
    };
  }
}

Envoy:

- name: envoy.filters.network.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          stat_prefix: grpc_json
          codec_type: AUTO
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains: ["*"]
              routes:
              # NOTE: by default, matching happens based on the gRPC route, and not on the incoming request path.
              # Reference: https://www.envoyproxy.io/docs/envoy/latest/configuration/http_filters/grpc_json_transcoder_filter#route-configs-for-transcoded-requests
              - match: { prefix: "/helloworld.Greeter" }
                route: { cluster: grpc, timeout: { seconds: 60 } }
          http_filters:
          - name: envoy.filters.http.grpc_json_transcoder
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.grpc_json_transcoder.v3.GrpcJsonTranscoder
              proto_descriptor: "protos/helloworld.pb"
              services: ["helloworld.Greeter"]
              print_options:
                add_whitespace: true
                always_print_primitive_fields: true
                always_print_enums_as_ints: false
                preserve_proto_field_names: false
          - name: envoy.filters.http.router

The most important thing here is:

services: ["helloworld.Greeter"] specifies the set of services this grpc json transcoder should have effect on.  Therefore, for any other services, and for any other urls which do NOT belong to helloworld.Greeter, the transcoder filter should pass through.

The commit 6ce641f with strict_http_request_validation = true changed this behavior and it is a breaking change.  If in the helloworld.pb there is another service, and if the service is not specified in services: [...], which can be completely legit use case, it is now with strict_http_request_validation = true, it will return 400 Bad Request.

**Comment by qiwzhang** [comment]
Well,  I still prefer to add a config to guide such behavior. 

```
enum TranscodingHeaderError {
    Path_mismatch = 0,
    Unknown_query_parameter,
};

    // map of error type to status_code. If defined, such error will be rejected with the status code
    // otherwise, it will be pass through
    map<TranscodingHeaderError,  StatusCode> error_reject_map = 13;
```

**Comment by nareddyt** [comment]
I am also unsure about this. Agreed that the strict request validation behavior is not intuitive for your use case. But there are other use cases where rejecting unknown paths is considered a feature.

For example, you can config the router to allow all paths by default. If you turn on `strict_http_request_validation=true`, the transcoder will automatically reject unknown paths. So now everytime you update the gRPC service definition, the transcoder will automatically pick up the changes. No need to manually modify the route configuration for every update - I consider this a feature.

```
     - name: envoy.filters.network.http_connection_manager
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
          stat_prefix: grpc_json
          codec_type: AUTO
          route_config:
            name: local_route
            virtual_hosts:
            - name: local_service
              domains: ["*"]
              routes:
              # >>> Allows ALL requests through, relies on transcoder to reject unknown paths. <<<
              - match: { prefix: "/" }
                route: { cluster: grpc, timeout: { seconds: 60 } }
          http_filters:
          - name: envoy.filters.http.grpc_json_transcoder
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.http.grpc_json_transcoder.v3.GrpcJsonTranscoder
              proto_descriptor: "protos/helloworld.pb"
              services: ["helloworld.Greeter"]
              # >>> Will only allow HTTP/JSON requests that are defined in the helloworld.Greeter service. <<<
              strict_http_request_validation: true
          - name: envoy.filters.http.router 
```

I don't know if anyone wants this use case, but it would be nice to support with an option similar to @qiwzhang comment. Though I think having the map `error_reject_map` is overkill. Bad path should be pass through or HTTP 404, bad query params should be pass through or HTTP 400. We can support his by just having a list. Or we can have separate fields for each behavior, similar to the `json_print_options`.

Also as we discussed offline, keep in mind you can use per-route configs with the transcoder to support your use case without this PR.

**Comment by tangyan** [comment]
> I am also unsure about this. Agreed that the strict request validation behavior is not intuitive for your use case. But there are other use cases where rejecting unknown paths is considered a feature.

After reading related bugs and discussions, I don't think this is really about our use case.

The key is, in previous existing behavior, grpc transcoder only touches/transforms requests when url paths match the urls defined in the services [..].  If the url doesn't match, it is then an unknown URL for the services [...].  The "unknown" URL cannot be decided if it is a wrong url or it's an intentional correct url which transcoder should NOT touch.

In other words, the newly added behavior effectively makes the services [...] not as meaningful as before.  This also means the existing envoy doc/example also needs to be changed.

**Comment by antoniovicente** [comment]
strict_http_request_validation seems to be a new config field.  Did the behavior of the proxy change when this field is set to false?

It is unclear to me if 1 config option is worse than multiple config options that provide more fine grained control about what requests are allow/disallowed.  1 config option seems simpler than multiple options.

Is there a more comprehensive writeup about problems experienced when doing json transcoding with strict_http_request_validation set to true?

**Comment by htuch** [comment]
Yeah, also unclear to me whether this is about the behavior before enabled the strict HTTP validation flag. @nareddyt @tangyan  did https://github.com/envoyproxy/envoy/pull/14715 change behaviors for anyone who didn't opt into the new flag?

**Comment by nareddyt** [comment]
No, it did not change any behavior for users who did not enable the option. I do not consider it a breaking change. I do understand the conceptual concern.

**Comment by htuch** [comment]
@nareddyt @qiwzhang @tangyan in that case, it would be great if you can agree on a design here. Given that the new fields landed only very recently, we can make breaking changes for 14 days since they first landed, see https://github.com/envoyproxy/envoy/blob/main/api/API_VERSIONING.md. So you have some flexibility here if you want to reconsider design.

**Comment by tangyan** [comment]
> No, it did not change any behavior for users who did not enable the option. I do not consider it a breaking change. I do understand the conceptual concern.

nareddyt is right on that if the option is not enabled, it is not a breaking change.  If the flag is turned on, then the behavior changed and grpc json trancoder becomes apply to all urls.

**Comment by tangyan** [comment]
> @nareddyt @qiwzhang @tangyan in that case, it would be great if you can agree on a design here. Given that the new fields landed only very recently, we can make breaking changes for 14 days since they first landed, see https://github.com/envoyproxy/envoy/blob/main/api/API_VERSIONING.md. So you have some flexibility here if you want to reconsider design.

We can figure this out.  Not a problem.

**Comment by nareddyt** [comment]
@tangyan do you have any concerns with introducing an enum to configure the behavior as described in my previous comment? The primary reason is that I want to keep the config extensible. If someone requests different behavior in the future, it will be a pain to deprecate and introduce new fields.

I can take over this PR if you are busy and get it fixed this week before the 14 day deadline.

**Comment by tangyan** [comment]
> @tangyan do you have any concerns with introducing an enum to configure the behavior as described in my previous comment? The primary reason is that I want to keep the config extensible. If someone requests different behavior in the future, it will be a pain to deprecate and introduce new fields.
> 
> I can take over this PR if you are busy and get it fixed this week before the 14 day deadline.

I have no concern, and would thank you if you can take the PR.  My main ask is that the behavior needs to be backward compatible without much surprise.  As long as it is configurable, it's fine, :).

Thanks!

**Comment by tangyan** [comment]
Close this in favor of #15019
